### ISSUE TITLE
Add end-of-life ("shutdown"? "dispose"? "close"?) semantics to grpc.Channel

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [grpc/grpc#12531]
Currently we use the Python interpreter's calling our `__del__` implementations to [shut down and destroy](https://github.com/grpc/grpc/blob/538418da961433e31d7545434b2d16bac4eb38d5/include/grpc/grpc.h#L285) the underlying gRPC Core resources associated with a `grpc.Channel`. Since there is no guarantee that memory is the only resource consumed by a `grpc.Channel` and since some garbage collectors only collect garbage when memory is scarce, there is a liability that applications might run out of those other resources (file descriptors and so on) when they are in fact perfectly reclaimable.

We should add an end-of-life method to `grpc.Channel`. We should probably also implement context manager semantics for both `grpc.Channel`s and `grpc.Server`s (we did so in both the Alpha and Beta gRPC Python APIs; what went poorly, I think, was making them _required_).

**Comment by dvaldivia** [comment]
+1

**Comment by aamitdb** [comment]
Hey, this is an example of the type of issue I'd like to have resolved by this ticket. 

1. Start a gRPC server e.g. https://github.com/grpc/grpc/blob/master/examples/python/route_guide/route_guide_server.py#L29
2. Spawn a daemon thread that is calling the server
```
import threading
import sys
import time

import grpc

import route_guide_pb2
import route_guide_pb2_grpc


def guide_list_features(stub):
    rectangle = route_guide_pb2.Rectangle(
        lo=route_guide_pb2.Point(latitude=400000000, longitude=-750000000),
        hi=route_guide_pb2.Point(latitude=420000000, longitude=-730000000))
    features = stub.ListFeatures(rectangle)
    for f in features:
        pass


def send_messages():
    channel = grpc.insecure_channel('localhost:50051')
    stub = route_guide_pb2_grpc.RouteGuideStub(channel)
    while True:
        guide_list_features(stub)


if __name__ == '__main__':
    t = threading.Thread(target=send_messages)
    t.daemon = True
    t.start()
    time.sleep(float(sys.argv[1]))
    print(time.time())
```
3. Run it in a loop
```
for i in {1..1000000};
do
python sample.py .2 >> /tmp/attempt.log 2>&1
done
```

Observer a variety of errors e.g.
1. `Exception in thread Thread-2 (most likely raised during interpreter shutdown):Exception in thread Thread-22 (most likely raised during interpreter shutdown):1517513380.4
1517513380.74`
2. 
```
Traceback (most recent call last):
   File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_innerException in thread Thread-2 (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 754, in run
  File "/tmp/grpc18/local/lib/python2.7/site-packages/grpc/_channel.py", line 816, in _poll_connectivity
  File "/usr/lib/python2.7/threading.py", line 168, in acquire
<type 'exceptions.TypeError'>: 'NoneType' object is not callable
```
3. Core dumps (can provide backtrace). 

I've played around with this example to try and cal `del` or `__del__` on the channel in an atexit handler, or even just open the channel itself and I can reproduce these errors.

I've tried on gRPC 1.8.4 and 1.9.0rc3, on both mac os (python 2.7.10) and Ubuntu 16.04.3 LTS VM (python 2.7.12)

**Comment by meawoppl** [comment]
Python `__del__` handlers are a mess, at the very least there should a call which allows the caller to close out connections/threads/etc left by the grpc backend.  I discovered this because our CI check for tests that leave dangling threads.  Simpy opening a insecure channel, and waiting for it to become live is enough to leave threads dangling.

**Comment by nathanielmanistaatgoogle** [comment]
I know this issue has fans (I'm the biggest one) so it is with disappointment that I have to say that resolution of this issue is going to slip from upcoming release `1.11` and will now be targeted to `1.12`.

Sorry friends.

**Comment by nathanielmanistaatgoogle** [comment]
I'm going to call this fixed as of the integration of [pull request 15254](https://github.com/grpc/grpc/pull/15254). I've got a list of follow-up items but I think this is done.

If you're working from source, please try it out.

**Comment by hubo1016** [comment]
@nathanielmanistaatgoogle It seems the new `close()` method is a blocking method. Is there any asynchronous way to close the channel? For example, add a `channel_close_future()` method to create a future for the channel shutdown. Or simply add a `start_close()` method and subscribe to the channel status.

This is useful in many ways, for example, shutdown multiple channels.

**Comment by nathanielmanistaatgoogle** [comment]
@hubo1016: how long is it blocking for you? It's not supposed to be a blocking method, but due to [issue 3064](https://github.com/grpc/grpc/issues/3064) it may block for up to 200ms. Are you seeing it block for longer than that?

**Comment by hubo1016** [comment]
@nathanielmanistaatgoogle no, I was just wondering, because I saw some wait operations in Cython code.

**Comment by hubo1016** [comment]
@nathanielmanistaatgoogle 
There may be an issue that after upgrading to grpc1.12.0, code that do not explicitly close the channel causes connection leak, makes this a break change. Maybe the new implementation creates reference loops and breaks the reference counting, so the connections may still be alive until GC collects them? I'm still tracing down this issue, but I would like to report.

**Comment by nathanielmanistaatgoogle** [comment]
@hubo1016: yes, it would be a problem if gRPC Python 1.12 leaked connections in a way that gRPC Python 1.11 did not leak connections. Have you progressed any further in your tracing down this question?

**Comment by meawoppl** [comment]
The tests in our codebase that exercised we're skipped for the last few months and have rotted.  I will be spending some cycles today trying to get them running again.

**Comment by meawoppl** [comment]
I am now encountering *segfaults*, this particular use case is monitored in a python thread to add to the tracing confusion. I am seeing if I can Smith a reproduction outside of that scope

**Comment by meawoppl** [comment]
Alright, I did some major digging and here are the things I learned:
- If you simply apply a version bump from `1.8.4` -> `1.12.0` the behaviour in Python subprocess/`fork()` changes subtly (this is under some integration tests in our stack)
- These changes are separate from other issued in the space most notable the #15334 

The core symptom I see is something like:
```
E0621 22:58:45.924080283   11368 server.cc:1265]             assertion failed: gpr_atm_acq_load(&server->shutdown_flag) || !server->listeners
Aborted
```

I can't get a very good trace because we use pytest's multiprocessing interface to run integration tests which get pretty hard-aborted killing the logging/tracebacks you might hopw for.  

I have a internal ticket we are going to work on to allow that to be run serially, but my suspicion looking at the line in question above is somewhere the shutdown flag is getting shared in an inappropriate way.  

Has a similar smell as this #14917 but I can't be certain on that one.
